contracts/FeralfileArtworkV4_1.sol

```diff
@@ -0,0 +1,184 @@
+    /// @param oldAddresses_ - the old addresses to replace
+    /// @param newAddresses_ - the new addresses to replace
+    function replaceAdvanceAddresses(
+        address[] calldata oldAddresses_,
+        address[] calldata newAddresses_
+    ) external onlyOwner {
+        if (oldAddresses_.length != newAddresses_.length) {
+            revert InvalidAdvanceAddressesAndAmounts();
+        }
+        for (uint256 i = 0; i < oldAddresses_.length; i++) {
+            if (newAddresses_[i] == address(0)) {
+                revert InvalidAdvanceAddress();
+            }
+            if (advances[newAddresses_[i]] > 0) {
+                revert AdvanceAddressAlreadyUsed();
+            }
+            advances[newAddresses_[i]] = advances[oldAddresses_[i]];
+            delete advances[oldAddresses_[i]];
+        }
+    }
+
+    /// @notice pay to get artworks to a destination address. The pricing, costs and other details is included in the saleData
+    /// @param r_ - part of signature for validating parameters integrity
+    /// @param s_ - part of signature for validating parameters integrity
+    /// @param v_ - part of signature for validating parameters integrity
+    /// @param saleData_ - the sale data
+    function buyArtworks(
+        bytes32 r_,
+        bytes32 s_,
+        uint8 v_,
+        SaleData calldata saleData_
+    ) external payable override {
+        require(_selling, "FeralfileExhibitionV4: sale is not started");
+        super._checkContractOwnedToken();
+        validateSaleData(saleData_);
+
+        saleData_.payByVaultContract
+            ? vault.payForSale(r_, s_, v_, saleData_)
+            : require(
+                saleData_.price == msg.value,
+                "FeralfileExhibitionV4: invalid payment amount"
+            );
+
+        bytes32 message = keccak256(
+            abi.encode(block.chainid, address(this), saleData_)
+        );
+
+        if (!isValidSignature(message, r_, s_, v_)) {
+            revert InvalidSignature();
+        }
+
+        uint256 itemRevenue;
+        if (saleData_.price > saleData_.cost) {
+            itemRevenue =
+                (saleData_.price - saleData_.cost) /
+                saleData_.tokenIds.length;
+        }
+
+        uint256 distributedRevenue;
+        uint256 platformRevenue;
+        for (uint256 i = 0; i < saleData_.tokenIds.length; i++) {
+            // send NFT
+            _safeTransfer(
+                address(this),
+                saleData_.destination,
+                saleData_.tokenIds[i],
+                ""
+            );
+            // distribute royalty
+            RevenueShare[] memory revenueShares = saleData_.revenueShares[i];
+            uint256 remainingRev = itemRevenue;
+
+            // deduct advances payment from revenue
+            for (uint256 j = 0; j < revenueShares.length; j++) {
+                if (
+                    advances[revenueShares[j].recipient] > 0 && remainingRev > 0
```

Comment:
This could be simplified 
```
for (uint256 j = 0; j < revenueShares.length && remainingRev > 0; j++) {
                uint256 remainingAdvanceAmount = advances[revenueShares[j].recipient];
                uint256 rev = remainingAdvanceAmount >= remainingRev ? remainingRev : remainingAdvanceAmount;
                platformRevenue += rev;
                advances[revenueShares[j].recipient] -= rev;
                remainingRev -= rev;
            }
```